[RFC] Add max_bytes_to_use_mmap_io

azat opened the pull request:
**NOTE: this patch requires, and includes for now, #23211 (submitted separately, since maybe #23211 will be backported)**

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add max_bytes_to_use_mmap_io

Detailed description / Documentation draft:
Since now (#23211) estimated memory that is used from mmap'ed region is accounted
in the memory tracker, we should have possibility to disable mmap if the
estimated memory is too big.

azat committed:
Add max_bytes_to_use_mmap_io

Since now estimated memory that is used from mmap'ed region is accounted
in the memory tracker, we should have possibility to disable mmap if the
estimated memory is too big.

azat committed:
Change max_bytes_to_use_mmap_io default to 1GiB

azat commented on src/Core/Settings.h:
```diff
@@ -143,6 +143,7 @@ class IColumn;
     \
     M(UInt64, min_bytes_to_use_direct_io, 0, "The minimum number of bytes for reading the data with O_DIRECT option during SELECT queries execution. 0 - disabled.", 0) \
     M(UInt64, min_bytes_to_use_mmap_io, (64 * 1024 * 1024), "The minimum number of bytes for reading the data with mmap option during SELECT queries execution. 0 - disabled.", 0) \
+    M(UInt64, max_bytes_to_use_mmap_io, 0, "The maximum number of bytes for reading the data with mmap option during SELECT queries execution. 0 - no limit.", 0) \
```
Note, that right now it is disabled, but I would rather set it 1GiB (or similar)

alexey-milovidov commented:
In my estimate it should shine on large reads.

alexey-milovidov commented:
See the MSan report in `OptimizeShardingKeyRewriteInVisitor.cpp:50`.

azat commented:
>In my estimate it should shine on large reads.

Agree, there are use cases when this should work well, something like reading all data from the one table in multiple queries, but  how much better it will? Do you have any numbers?

azat commented:
>See the MSan report in OptimizeShardingKeyRewriteInVisitor.cpp:50.
>[AST fuzzer (MSan) — SUMMARY: MemorySanitizer: use-of-uninitialized-value obj-x86_64-linux-gnu/../src/Interpreters/OptimizeShardingKeyRewriteInVisitor.cpp:50:28](https://clickhouse-test-reports.s3.yandex.net/23212/d02eb3f6b7a8dc07bbed5605e1afcf7634436b3a/fuzzer_msan/report.html#fail1)

#23219

azat commented:
>[Functional stateless tests flaky check (address) — fail: 1, passed: 199](https://clickhouse-test-reports.s3.yandex.net/23212/d02eb3f6b7a8dc07bbed5605e1afcf7634436b3a/functional_stateless_tests_flaky_check_(address).html#fail1)

Will mark `01818_min_bytes_to_use_mmap_io_memory_tracking` as long if the idea of the patches looks good.

alexey-milovidov commented:
> but how much better it will? Do you have any numbers?

Yes, it's in perf tests in a PR when it was enabled.

alexey-milovidov commented:
#22206 #22326

azat committed:
Mark 01818_min_bytes_to_use_mmap_io_memory_tracking as long

https://clickhouse-test-reports.s3.yandex.net/23212/d02eb3f6b7a8dc07bbed5605e1afcf7634436b3a/functional_stateless_tests_flaky_check_(address).html#fail1

azat commented:
>In my estimate it should shine on large reads.

I'm fine with changing the default to any sensible value (1-10GB), but after #23211 it is better to have max limit for mmap.

>#22206

Seems that hits_none created from hits (in 00162_mmap_compression_none.sql) and hits that I have is:

```
65M     hits/data/datasets/hits/201403_1_9_3/WatchID.bin
68M     hits_v1/data/datasets/hits_v1/201403_10_18_2/WatchID.bin
767M    hits_100m_obfuscated_v1/data/default/hits_100m_obfuscated/201307_1_96_4/WatchID.bin
```

So everything is less then 1G.

azat committed:
performance/mmap_io: run w/o mmap too (for comparison)

alexey-milovidov commented:
I cannot imagine what max limit should be taking into account the fact that min threshold is 1 GB and 150 GB is the max part size.

alexey-milovidov commented:
Do you have any motivating use cases?

azat commented:
The motivation was to provide a knob to disable mmap for big parts, since with multiple queries that just reads those parts the queries may fail due to memory limit exceeded (for total memory tracker), after #23211 had been merged.

azat commented:
Also note that after #23211 merged perf tests started to fail with memory limit exceeded error (but for query), I will fix this, but let's decide with this first.

Do you think that there is no sense in this `max_bytes_to_use_mmap_io` knob?

alexey-milovidov commented:
Either I don't understand the motivation or there is very little sense to add this feature.

> with multiple queries that just reads those parts the queries may fail due to memory limit exceeded (for total memory tracker)

Does not look as an issue. Memory is calculated in some reasonable way (even if it is calculated multiple times for shared mappings, it is still reasonable for memory accounting). The user will get an error about memory limit maybe slightly sooner than expected.

azat commented:
>Does not look as an issue. Memory is calculated in some reasonable way (even if it is calculated multiple times for shared mappings, it is still reasonable for memory accounting). The user will get an error about memory limit maybe slightly sooner than expected.

Suppose you have the following pattern:
- 50-100 small queries (that reads small amount of data), that are running every second
- and 1 query that reads all the table (big table), that you need to run daily

So when you will execute the query that reads everything from the table, the server will mmap lots of files and small queries will start to fail, because of the total memory limit, this know will allow to limit size of files that should be mmapped.

Also [I did some testing with fio](https://gist.github.com/azat/3d6c8d82bdd91e7a38d997fd6bcfd574) for read/read with page cache/read direct/read mmap and I don't see definite win for using mmap with big enough files, since page-cache + readahead should work quite well. So I think that this should be good enough motivation, thoughts?

I don't object that there can be some use case when it will give some benefits, but I doubt that it is real production load (at least now), and if someone will face it, the max_bytes_to_use_mmap_io maybe set to 0.

akuzm commented:
Why are mmapped files counted towards memory usage at all? This looks like the core problem. mmapping a file doesn't use up the available memory (RAM+swap).

alexey-milovidov commented:
> Why are mmapped files counted towards memory usage at all? This looks like the core problem. mmapping a file doesn't use up the available memory (RAM+swap).

I agree. It was in previous PR that is already reverted.
